MWPW-168589: Placeholders in M@S

Axelcureno opened the pull request:
This PR introduces an all-new Placeholder management experience in Merch at Scale Studio that enables CRUD operations using Odin as the data source.
Other improvements included in this PR are:

- code enhancements and cleanup.
- Refactored, more resilient and consistent routing navigation.
- Fixes and expands test coverage, including a new mock service that covers more files.
- Lint fixes.

Resolve  https://jira.corp.adobe.com/browse/MWPW-168589

Test URLs:
- Before: https://main--mas--adobecom.aem.live/studio.html?page=placeholders#path=nala&locale=en_US
- After: https://mwpw-168589--mas--adobecom.aem.live/studio.html?page=placeholders#path=nala&locale=en_US

yesil commented on studio/src/mas-side-nav.js:
```diff
@@ -77,19 +81,17 @@ class MasSideNav extends LitElement {
                     value="home"
                     @click="${navigateToPage('welcome')}"
-                    selected
+                    ?selected=${this.currentPage.value === 'welcome'}
```
Please use `Store.page`. We discussed pros & cons of using StoreController for accessing data and the outcome was to use `Store.` api directly. 
Thus, we keep also automatic code completion.

Axelcureno replied:
Addressed

yesil commented on studio/src/mas-side-nav.js:
```diff
@@ -77,19 +81,17 @@ class MasSideNav extends LitElement {
+                <sp-sidenav-item
+                    label="Placeholders"
+                    value="placeholders"
+                    @click="${navigateToPage('placeholders')}"
```
time to make these values (placeholders, content, welcome) a constant?

Axelcureno replied:
done

yesil commented on studio/src/store.js:
```diff
@@ -97,105 +109,94 @@ export function navigateToPage(value) {
+                  ? defaultValue[key]
+                  : defaultValue;
+
+            if (defaultForKey !== undefined) {
```
I also changed this part in my PR #196, I wonder what was the reason you changed this?
in mine I noticed, that `syncFromHash` was causing extra search calls, also breaking the loading indicator.

Axelcureno replied:
In general I made the handling of search params more robust by implementing standardized URL parsing, supports complex data structures via JSON serialization and enhances maintainability with cleaner, more efficient code organization. This particular section improves performance by avoiding multiple reactive updates when setting several default values at once (like the issue you described).

yesil commented on studio/src/store.js:
```diff
@@ -97,105 +109,94 @@ export function navigateToPage(value) {
 
-    // Initialize
-    syncFromHash();
+    store.subscribe((value) => {
```
❤️

yesil commented on studio/src/store.js:
```diff
@@ -209,11 +210,47 @@ export function unlinkStoreFromHash(store) {
         Store.page.set('content');
 });
+
+document.addEventListener('DOMContentLoaded', () => {
```
store.js is growing too fast.
can we move this in a separate file?
name suggestions: router.js, deeplink.js.

Axelcureno replied:
I used your suggestion `router.js` and moved navigation-related stuff there.

yesil commented on studio/src/studio.js:
```diff
@@ -52,10 +53,24 @@ class MasStudio extends LitElement {
+        );
+    }
+
+    updatePageFromUrl() {
```
initialization of the store from the URL should be centralized at a single place. Otherwise it will be very hard to troubleshoot.

yesil commented on studio/test/aem/aem.test.js:
```diff
@@ -1,8 +1,106 @@
+    });
+
+    describe('fragments.create', () => {
+        it('should create fragment successfully', async () => {
```
I don't think we can test the creation.

yesil commented on studio/test/aem/aem.test.js:
```diff
@@ -1,8 +1,106 @@
+    });
+
+    describe('fragments.publish', () => {
+        it('should publish fragment successfully', async () => {
```
to test aem.js, I recommend to test the payload of the fetch requests being send.

yesil commented on studio/test/aem/mas-placeholders.test.js:
```diff
@@ -0,0 +1,300 @@
+                {
+                    key: 'test-key',
+                    value: 'test-value',
+                    _fragment: {
```
please do not start variables with `_`.

yesil commented on studio/test/aem/mas-placeholders.test.js:
```diff
@@ -0,0 +1,300 @@
+         * Mock Store object that simulates the reactive data store
+         * Includes placeholders, search, filters and page state
+         */
+        window.Store = {
```
maybe you can create a store.js in mocks folder?

yesil commented on studio/test/aem/mas-placeholders.test.js:
```diff
@@ -0,0 +1,300 @@
+                this.editingPlaceholder = null;
+
+                // Emit success toast
+                window.Events.toast.emit({
```
the events such as `placeholderAdded` do not seem to be leveraged here.

yesil commented on studio/test/mocks/aem-service-mock.js:
```diff
@@ -0,0 +1,138 @@
+    // Define fragment models
+    this._fragmentModels = new Map();
+    this._fragmentModels.set('test-model', {
+      id: 'test-model',
```
please do not start variables with underscore.

yesil commented on studio/test/mocks/aem-service-mock.js:
```diff
@@ -0,0 +1,138 @@
+    
+    return method;
+  }
+} 
```
EOF

yesil commented on studio/test/mocks/aem-service-mock.js:
mocks should have the same name as the module they are mocking for ease of discoverability.

yesil commented on studio/test/mocks/aem-service-mock.js:
```diff
@@ -0,0 +1,138 @@
+/**
+ * Complete AEM service mock for testing
+ */
+export class AEMServiceMock {
```
where is this used? I would like to avoid reviewing AI generated code that is not used.

Axelcureno replied:
Sorry, I forgot to remove this.

yesil commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,1479 @@
+}
+
+class MasPlaceholders extends LitElement {
+    static styles = css`
```
can you move styles to a css.js file?  when there are too many styles, it is hard to edit the file.

yesil commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,1479 @@
+        };
+
+        this.selectedFolder = new StoreController(
+            this,
```
can you use a single `ReactiveController` instead?

reactiveController = new ReactiveController(this, [...stores])

yesil commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,1479 @@
+        if (masRepository) {
+            if (this.selectedFolder.value?.path) {
+                Store.placeholders.list.loading.set(true);
+                setTimeout(() => {
```
why setTimeout is needed?

Axelcureno replied:
removed

yesil commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,1479 @@
+                key: this.editedKey,
+                value: this.editedValue,
+                status: savedFragment.status || 'Draft',
+                _fragment: savedFragment,
```
please respect coding convention, we do not start variables with underscore.

Axelcureno replied:
fixed

yesil commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,1479 @@
+            );
+            Store.placeholders.list.data.set(updatedPlaceholders);
+
+            setTimeout(() => {
```
why `setTimeout` is needed?

Axelcureno replied:
removed

yesil commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,1479 @@
+    }
+
+    handleExport() {
+        // ToDo: Implement export functionality
```
is this needed?

Axelcureno replied:
It will in the future but I have removed it in the meantime. You may noticed I also removed other stuff like this that is not used.

yesil commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,1479 @@
+            this.repository.searchPlaceholders();
+        }
+
+        this.requestUpdate();
```
nit: this should not be needed, as it will refresh when one of the observed stores is updated.

Axelcureno replied:
Removed

st-angelo-adobe commented on studio/src/store.js:
```diff
@@ -8,7 +8,11 @@ import {
     setHashParams,
```
There are some unused methods being imported

st-angelo-adobe commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,1479 @@
+        );
+        this.selectedLocale = new StoreController(
+            this,
+            Store.filters?.locale || new ReactiveStore('en_US'),
```
We shouldn't be creating new reactive stores anywhere outside the `store.js` file. Instead we should use the existing one, so we simply subscribe to the `filters` store, and to the `folders.data` store.

Axelcureno replied:
🫠 removed

st-angelo-adobe commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,1479 @@
+
+    handleLocaleChange(event) {
+        const newLocale = event.detail.locale;
+        Store.filters.value.locale = newLocale;
```
We should NEVER manually set a store's value like this, because then the subscribers don't get notified.
We should only update the value of the store using `set`.

Axelcureno commented on studio/src/store.js:
```diff
@@ -97,105 +109,94 @@ export function navigateToPage(value) {
-    // Subscribe
-    window.addEventListener('hashchange', syncFromHash);
-    store.subscribe(syncToHash);
-    store.setMeta('hashLink', { from: syncFromHash, to: syncToHash });
```
I revamped the URL parameter handling mechanism.
The new implementation delivers significant advantages through five key improvements:
1. **Better Data Structure Support**: Properly handles complex objects and arrays through JSON serialization, ensuring accurate preservation of user settings in URLs.
2. **Performance Optimization**: Collects all parameter changes and applies them in a single update, eliminating unnecessary re-renders and improving page load speed.
3. **Smarter History Management**: Uses the History API to update the current browser history entry rather than creating multiple entries, providing a more intuitive navigation experience.
4. **Consistent Default Values**: Implements a clear, predictable system for applying default values when parameters are missing, ensuring a reliable user experience.
5. **Enhanced Shareability**: Makes URLs perfectly capture application state, allowing users to bookmark specific views or share exact configurations with colleagues.

yesil reviewed (commented):
reset filters button is broken.

<img width="145" alt="image" src="https://github.com/user-attachments/assets/1185a54d-578d-4b32-8c39-fec302d8e432" />

Axelcureno commented on studio/src/events.js:
```diff
@@ -3,6 +3,10 @@ import MasEvent from './reactivity/mas-event.js';
 const Events = {
     toast: new MasEvent(),
     fragmentAdded: new MasEvent(),
+    placeholderAdded: new MasEvent(),
```
removed

Axelcureno commented:
> reset filters button is broken.
> 
> <img alt="image" width="145" src="https://private-user-images.githubusercontent.com/330057/425536836-1185a54d-578d-4b32-8c39-fec302d8e432.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI1Nzk3OTMsIm5iZiI6MTc0MjU3OTQ5MywicGF0aCI6Ii8zMzAwNTcvNDI1NTM2ODM2LTExODVhNTRkLTU3OGQtNGIzMi04YzM5LWZlYzMwMmQ4ZTQzMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyMVQxNzUxMzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YTgzY2NjOTIxY2EzMGE3MjlkOWU3YzFhZDhmOWJlNWQ1NWM2MDIwYTkzNzY2ZDgyNGMzMTVmYWM0Nzg0ZjY5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.JEO7vi-QbsbHR4cQCR3La2ZO2WUhUjUPzaahCh1FvbU">

@yesil good callout, thanks. I have fixed it by doing the following:
- Made sure both data stores (search and filter) are reset when clicking "reset filters" (previously only one was being updated).
- Added code to explicitly clear filter parameters from the URL (this was causing the issue).
- Improved how the app detects filter changes by adding special handling for empty tag arrays and **subscribing to filter changes directly**.

3ch023 reviewed (changes requested):
Hi @Axelcureno , was checking it today and it looks great!
one small issue i found is that 'index' placeholder containing all other placeholders is not updated when i add a new placeholder.
@npeltier can help with details on how its supposed to work, could you sync with him pls and address?
this is my Milo test page with M@S card with 2 placeholders:
https://stage--milo--adobecom.aem.page/drafts/mariia/document5

studio: https://mas.adobe.com/studio.html#path=sandbox&query=a61c723c-3c07-43d5-a082-97609a4da4e7&page=content
you can see {{cai-default}} works, but my new {{mariia-placeholder}} not yet.
i assume because https://odin.adobe.com/adobe/sites/fragments?path=/content/dam/mas/sandbox/en_US/dictionary/index doesn't have a reference to it.

+ i wouldn't check Nala till Ilyases PR on collection goes in. once its merged you can pull latest & recheck

npeltier reviewed (changes requested):
Hi @Axelcureno , as discussed offline, when a new placeholder is created for a surface / locale, and saved (before it's even published) you should  
1. check for existence of an index fragment for that surface / locale if, not existing create it,
2. add that new fragment as reference to index
3. publish index (that should publish new fragment too)

i discussed the drawback of having a draft published, but at that stage, the PH is not used so it's good.

Axelcureno commented on studio/src/mas-side-nav.js:
```diff
@@ -76,20 +81,18 @@ class MasSideNav extends LitElement {
-                    <sp-icon-graph-bar-vertical
-                        slot="icon"
-                    ></sp-icon-graph-bar-vertical>
+                <sp-sidenav-item
```
added

Axelcureno commented on studio/src/mas-top-nav.js:
```diff
@@ -3,9 +3,16 @@ import { LitElement, html, css, until } from 'lit';
 import Store from './store.js';
```
Which changes?

Axelcureno commented:
Addressed all your comments @npeltier @3ch023 :)

npeltier commented:
Hey @Axelcureno current published index still only have 4 entries, see: 
https://odin.adobe.com/adobe/sites/fragments/4218bdb3-8d25-4706-974d-21b1f0f13aa9

npeltier commented:
i tried to delete / create / publish and it's still the same

afmicka commented:
@Axelcureno Nala is not able to authenticate because something has changed in the studio url.
As far as i can see you are adding and placing `page=welcome` before the hash (`#`). Also moving `page=content` before it.
On your branch it is `?page=content#path=acom`, and before it was `#page=content&path=acom`. Is that intentional? 
Is yes, some of the test specs and authentication need to be updated to match it and be able to execute the tests.

npeltier commented:
also @Axelcureno FYI just tested creation of placeholder in an empty dictionary folder use case that didn't work. Not sure we make this part of your PR to be fixed (i think i saw create if not exist) or make it part of initial content script creation (we _will_ need one as placeholder needs permissions setup you won't do in your code). But either way change is requested here too:
- fix the create if not exist,
- remove that code
cc @3ch023 @markovicpredrag1209 @yesil

npeltier commented on studio/src/mas-repository.js:
```diff
@@ -314,36 +339,57 @@ export class MasRepository extends LitElement {
-                variant: 'info',
-                content: 'Creating fragment...',
-            });
+            const isPlaceholder = fragmentData.data && (fragmentData.data.key !== undefined || 
+                                                       fragmentData.parentPath?.includes('/dictionary/'));
+            this.showToast(isPlaceholder ? 'Creating placeholder...' : 'Creating fragment...');
```
i think you can just leave it as create Fragment, that's fine

Axelcureno replied:
done

npeltier commented on studio/src/mas-repository.js:
```diff
@@ -355,35 +401,89 @@ export class MasRepository extends LitElement {
+            throw new Error('No fragment provided for saving');
+        }
+        
+        const isDictionaryFragment = fragmentToSave.path?.includes('/dictionary/');
+        this.showToast(isDictionaryFragment ? 'Saving placeholder...' : 'Saving fragment...');
```
we shouldn't do that either, just saving frgment is fine

Axelcureno replied:
done

npeltier commented on studio/src/mas-repository.js:
```diff
@@ -355,35 +401,89 @@ export class MasRepository extends LitElement {
+            this.showToast(
+                isDictionaryFragment ? 'Placeholder successfully saved.' : 'Fragment successfully saved.',
+                'positive'
             );
```
please move this all to a different "updatePlaceholderIndex" method you call from placeholder js together within createion code, and fix the payload, right now it's set to fragment id, hence the publish is broken atm

Axelcureno replied:
done

Axelcureno commented on studio/src/mas-repository.js:
```diff
@@ -492,6 +500,791 @@ export class MasRepository extends LitElement {
+     * @param {Object} fragmentData - The fragment data to create
+     * @returns {Promise<Object>} - The created fragment
+     */
+    async createDictionaryFragment(fragmentData) {
```
done

Axelcureno commented on studio/src/mas-repository.js:
```diff
@@ -492,6 +500,791 @@ export class MasRepository extends LitElement {
+            
+            const newFragment = await this.createDictionaryFragment(fragmentObjectWithStatus);
+            
+            const publishedFragment = await this.publishDictionaryFragment(newFragment);
```
done

Axelcureno commented on studio/src/constants.js:
```diff
@@ -110,12 +111,22 @@ export const LOCALE_DEFAULT = 'en_US';
 export const TAG_STATUS_DRAFT_PATH = '/content/cq:tags/mas/status/draft';
 
+export const ROOT_PATH = '/content/dam/mas';
+export const DICTIONARY_MODEL_ID = 'L2NvbmYvbWFzL3NldHRpbmdzL2RhbS9jZm0vbW9kZWxzL2RpY3Rpb25uYXJ5';
```
it is being used.

Axelcureno commented on studio/src/mas-repository.js:
```diff
@@ -314,36 +339,57 @@ export class MasRepository extends LitElement {
             return new FragmentStore(fragment);
         } catch (error) {
-            this.processError(error, 'Failed to create fragment.');
+            this.processError(error, isPlaceholder ? 'Failed to create placeholder.' : 'Failed to create fragment.');
```
done

Axelcureno commented:
All remaining comments addressed @npeltier

afmicka commented on nala/studio/studio.spec.js:
```diff
@@ -34,13 +34,14 @@ export default {
                 cardid: '0bf35134-e5e4-4664-88d9-4b78203bf625',
             },
-            browserParams: '#path=nala&page=content',
+            browserParams: '?locale=en_US#path=nala&page=content',
```
also here. It is expected that the locale automatically gets recognized if user specifies only path or page

afmicka commented on nala/studio/studio.test.js:
```diff
@@ -127,7 +127,7 @@ test.describe('M@S Studio feature test suite', () => {
         baseURL,
     }) => {
-        const testPage = `${baseURL}${features[4].path}${miloLibs}`;
+        const testPage = `${baseURL}${features[4].path}${miloLibs}${features[4].browserParams}`;
```
i believe adding locale param here is also not needed. It you have issues with the tests without it, lets connect and discuss their behaviour.

npeltier reviewed (approved):
ok, works for me :-) this is great, thanks!

i'm still not a big fan of if tests within generic methods, but i guess this can be changed later

Axelcureno commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,1796 @@
+            const fragmentId = fragmentData.id;
+            const fragmentEndpoint = `${repository.aem.cfFragmentsUrl}/${fragmentId}`;
+
+            const getResponse = await fetch(fragmentEndpoint, {
```
It was initially in mas-repository but @npeltier asked me to move things here 🥲

Axelcureno commented:
@Roycethan my notes regarding testing of placeholders.
1. I have removed offer selector tool from placeholder creation, that is not in scope for now.
2. I was able to modify links on an existing placeholder (added/removed link, changed style) with no issues, please try again and if the issue persists let me know what placeholder you are trying with.
3. Please test only in locales that already have placeholders for now. This is happening because not all locales have the folder created in Odin. @npeltier said this will be handled separately in another ticket, a script to create the folders in Odin is needed.
4. This ticket is only for creating/modifying/deleting placeholders. No need to test placeholders resolving in cards for now (separate ticket exists for this use case).
5. Don't use Figma to compare the design, the current version is work in progress and has many things that we decided are not needed and should be removed (including status column). When a placeholder is created it is automatically published making "status" trivial.

Axelcureno commented:
I have addressed all of your last comments @yesil

Roycethan commented:
> @Roycethan my notes regarding testing of placeholders.
> 
> 1. I have removed offer selector tool from placeholder creation, that is not in scope for now.
> 2. I was able to modify links on an existing placeholder (added/removed link, changed style) with no issues, please try again and if the issue persists let me know what placeholder you are trying with.
> 3. Please test only in locales that already have placeholders for now. This is happening because not all locales have the folder created in Odin. @npeltier said this will be handled separately in another ticket, a script to create the folders in Odin is needed.
> 4. This ticket is only for creating/modifying/deleting placeholders. No need to test placeholders resolving in cards for now (separate ticket exists for this use case).
> 5. Don't use Figma to compare the design, the current version is work in progress and has many things that we decided are not needed and should be removed (including status column). When a placeholder is created it is automatically published making "status" trivial.

@Axelcureno  For edit link, try removing link from the below placeholder ( by clicking Remove link) and save (at right) and see link still exists:
<img width="1238" alt="image" src="https://github.com/user-attachments/assets/f9013de4-559f-4ab3-965b-bbe93125865f" />

yesil commented on studio/src/mas-fragment.js:
```diff
@@ -87,4 +87,4 @@ class MasFragment extends LitElement {
 }
 
-customElements.define('mas-fragment', MasFragment);
+customElements.define('mas-fragment', MasFragment);
```
EOF

yesil commented on studio/src/mas-content.js:
```diff
@@ -27,11 +28,32 @@ class MasContent extends LitElement {
         super.disconnectedCallback();
         Events.fragmentAdded.unsubscribe(this.goToFragment);
+
+        if (this.subscriptions && this.subscriptions.length) {
```
please prefer ReactiveController for pure refresh purposes.

yesil commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,2068 @@
+        this.reactiveController = new ReactiveController(this, [
+            Store.search,
+            Store.filters,
+            Store.folders?.data,
```
folders is always defined. 
`property?` should be used only for props that can be set/unset later.

yesil commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,2068 @@
+    static styles = styles;
+
+    static properties = {
+        searchQuery: { type: String, state: true },
```
👍

yesil commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,2068 @@
+                }
+            });
+
+            const saveResponse = await fetch(fragmentEndpoint, {
```
why not use, saveFragment from `mas-repository`

yesil commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,2068 @@
+
+        try {
+            const repository = this.repository;
+            if (!repository) {
```
I think this should be used only in the render method. 
All the dependencies that are needed, should be pre-conditions to render placeholders page.
Otherwise, this can of checks make the code only more verbose.

```
            if (!repository) {
                return false;
            }
```

yesil commented on studio/src/mas-repository.js:
```diff
@@ -90,6 +88,10 @@ export class MasRepository extends LitElement {
         this.style.display = 'none';
     }
 
+    disconnectedCallback() {
```
this is not needed.

yesil commented on studio/src/mas-placeholders.js:
```diff
@@ -0,0 +1,2068 @@
+            } catch (error) {
+                if (error.message?.includes('412')) {
+                    await new Promise((resolve) => setTimeout(resolve, 500));
+                    return await this.updateIndexFragment(
```
this can cause a loop forever.

yesil reviewed (commented):
no blocker comments, but certains things I already commented like using fetch only from aem.js, is still not fully implemented. I'm ok to merge if no regression though.
those can be addressed later.

Roycethan commented:
> > @Roycethan my notes regarding testing of placeholders.
> > 
> > 1. I have removed offer selector tool from placeholder creation, that is not in scope for now.
> > 2. I was able to modify links on an existing placeholder (added/removed link, changed style) with no issues, please try again and if the issue persists let me know what placeholder you are trying with.
> > 3. Please test only in locales that already have placeholders for now. This is happening because not all locales have the folder created in Odin. @npeltier said this will be handled separately in another ticket, a script to create the folders in Odin is needed.
> > 4. This ticket is only for creating/modifying/deleting placeholders. No need to test placeholders resolving in cards for now (separate ticket exists for this use case).
> > 5. Don't use Figma to compare the design, the current version is work in progress and has many things that we decided are not needed and should be removed (including status column). When a placeholder is created it is automatically published making "status" trivial.
> 
> @Axelcureno For edit link, try removing link from the below placeholder ( by clicking Remove link) and save (at right) and see link still exists: <img alt="image" width="1238" src="https://private-user-images.githubusercontent.com/104928176/429677716-f9013de4-559f-4ab3-965b-bbe93125865f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM3MjQxMTAsIm5iZiI6MTc0MzcyMzgxMCwicGF0aCI6Ii8xMDQ5MjgxNzYvNDI5Njc3NzE2LWY5MDEzZGU0LTU1OWYtNGFiMy05NjViLWJiZTkzMTI1ODY1Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQwM1QyMzQzMzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yYzdjZGRhNThmYzA2MTQ4ZmRlNWNlYzhiMTAxYzdhOGVlMjNiZjg3YzhhNzRkZGY3MjhhMWU1NmExODljZThkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.GBSs7pu0J0y-PFHcvhH2_kSJMSZxJaOJDQo9Fmvlly0">
> 

Also the text boxes can be made in line in one row 
<img width="1238" alt="image" src="https://github.com/user-attachments/assets/2bb13b2c-cacf-4999-a08a-de8a0c9fe838" />
